hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSNamesystem.java

```diff
@@ -3470,6 +3471,16 @@ HdfsFileStatus getFileInfo(final String src, boolean resolveLink,
       logAuditEvent(false, operationName, src);
       throw e;
     }
+    if (needLocation && haEnabled && haContext != null &&
+        haContext.getState().getServiceState() == OBSERVER &&
+        stat instanceof HdfsLocatedFileStatus) {
+      LocatedBlocks lbs = ((HdfsLocatedFileStatus) stat).getLocatedBlocks();
+      for (LocatedBlock b : lbs.getLocatedBlocks()) {
+        if (b.getLocations() == null || b.getLocations().length == 0) {
+          throw new ObserverRetryOnActiveException("Zero blocklocations for " + src);
+        }
+      }
+    }
```

Comment:
Can we pull this into a common method like:
```java
private void checkBlockLocationsIfObserver(Iterator<LocatedBlocks> blocksIter) throws ObserverRetryOnActive {
  if (haEnabled && haContext != null && haContext.getState().getServiceState() == OBSERVER) {
    ...
  }
}
```

or two methods like
```java
private boolean isObserver() { return haEnabled && haContext != null && haContext.getState().getServiceState() == OBSERVER; }
private void checkBlockLocationsForObserver(LocatedBlocks blocks) throws ObserverRetryOnActive { ... }
```

Point being that we have 3 places with almost identical logic here, we should try to consolidate.

Reply:
> Can we pull this into a common method like:
> 
> ```java
> private void checkBlockLocationsIfObserver(Iterator<LocatedBlocks> blocksIter) throws ObserverRetryOnActive {
>   if (haEnabled && haContext != null && haContext.getState().getServiceState() == OBSERVER) {
>     ...
>   }
> }
> ```
> 
> or two methods like
> 
> ```java
> private boolean isObserver() { return haEnabled && haContext != null && haContext.getState().getServiceState() == OBSERVER; }
> private void checkBlockLocationsForObserver(LocatedBlocks blocks) throws ObserverRetryOnActive { ... }
> ```
> 
> Point being that we have 3 places with almost identical logic here, we should try to consolidate.

good idea!